File: tests/java/test-app/src/main/java/com/godaddy/asherah/testapp/regression/EncryptMetastoreInteractions.java
package com.godaddy.asherah.testapp.regression;

import com.godaddy.asherah.testapp.testhelpers.KeyState;

class EncryptMetastoreInteractions {

  private KeyState cacheIK;
  private KeyState metaIK;
  private KeyState cacheSK;
  private KeyState metaSK;

  protected EncryptMetastoreInteractions(final KeyState cacheIK, final KeyState metaIK,
                                         final KeyState cacheSK, final KeyState metaSK) {

    this.cacheIK = cacheIK;
    this.metaIK = metaIK;
    this.cacheSK = cacheSK;
    this.metaSK = metaSK;
  }

  protected boolean shouldStoreIK() {

    if (cacheIK == KeyState.VALID) {
      return false;
    }

    if (metaIK != KeyState.VALID) {
      return true;
    }

    // At this stage IK is valid in metastore.
    // The existing IK can only be used if the SK is valid in cache,
    if (cacheSK == KeyState.VALID) {
      return false;
    }
    // or if the SK is missing from the cache but is valid in metastore.
    else if (cacheSK == KeyState.EMPTY) {
      if (metaSK == KeyState.VALID) {
        return false;
      }
    }

    return true;
  }

  protected boolean shouldStoreSK() {

    if (cacheIK == KeyState.VALID) {
      return false;
    }

    return cacheSK != KeyState.VALID && metaSK != KeyState.VALID;
  }

  protected boolean shouldLoadSK() {

    if (cacheIK == KeyState.VALID) {
      return false;
    }

    if (metaIK != KeyState.VALID) {
      return false;
    }

    if (cacheSK == KeyState.EMPTY) {
      return true;
    }

    return false;
  }

  protected boolean shouldLoadLatestIK() {
    return cacheIK != KeyState.VALID;
  }

  protected boolean shouldLoadLatestSK() {

    if (cacheIK == KeyState.VALID) {
      return false;
    }

    if (metaIK == KeyState.VALID) {
      // Because cacheSK points to a retired and latest value in cache,
      // we need to loadLatest SK from metastore
      if (cacheSK == KeyState.RETIRED) {
        return true;
      }

      // We know it's not in the cache, so can we need to load the latest SK in metastore
      return cacheSK == KeyState.EMPTY && metaSK != KeyState.VALID;
    }

    return cacheSK != KeyState.VALID;
  }
}

File: languages/csharp/AppEncryption/AppEncryption.IntegrationTests/Regression/EncryptMetastoreInteractions.cs
using GoDaddy.Asherah.AppEncryption.IntegrationTests.TestHelpers;

namespace GoDaddy.Asherah.AppEncryption.IntegrationTests.Regression
{
    public class EncryptMetastoreInteractions
    {
        private readonly KeyState cacheIK;
        private readonly KeyState metaIK;
        private readonly KeyState cacheSK;
        private readonly KeyState metaSK;

        protected internal EncryptMetastoreInteractions(
            KeyState cacheIK, KeyState metaIK, KeyState cacheSK, KeyState metaSK)
        {
            this.cacheIK = cacheIK;
            this.metaIK = metaIK;
            this.cacheSK = cacheSK;
            this.metaSK = metaSK;
        }

        protected internal bool ShouldStoreIK()
        {
            if (cacheIK == KeyState.Valid)
            {
                return false;
            }

            if (metaIK != KeyState.Valid)
            {
                return true;
            }

            // At this stage IK is valid in metastore
            // The existing IK can only be used if the SK is valid in cache
            // or if the SK is missing from the cache but is valid in metastore
            if (cacheSK == KeyState.Valid)
            {
                return false;
            }
            else if (cacheSK == KeyState.Empty)
            {
                if (metaSK == KeyState.Valid)
                {
                    return false;
                }
            }

            return true;
        }

        protected internal bool ShouldStoreSK()
        {
            if (cacheIK == KeyState.Valid)
            {
                return false;
            }

            return cacheSK != KeyState.Valid && metaSK != KeyState.Valid;
        }

        protected internal bool ShouldLoadSK()
        {
            if (cacheIK == KeyState.Valid)
            {
                return false;
            }

            if (metaIK != KeyState.Valid)
            {
                return false;
            }

            if (cacheSK == KeyState.Empty)
            {
                return true;
            }

            return false;
        }

        protected internal bool ShouldLoadLatestIK()
        {
            return cacheIK != KeyState.Valid;
        }

        protected internal bool ShouldLoadLatestSK()
        {
            if (cacheIK == KeyState.Valid)
            {
                return false;
            }

            if (metaIK == KeyState.Valid)
            {
                // Because cacheSK points to a retired and latest value in cache,
                // we need to loadLatest SK from metastore
                if (cacheSK == KeyState.Retired)
                {
                    return true;
                }

                // We know it's not in the cache, so can we need to load the latest SK in metastore
                return cacheSK == KeyState.Empty && metaSK != KeyState.Valid;
            }

            return cacheSK != KeyState.Valid;
        }
    }
}
